csharp/ql/src/semmle/code/csharp/exprs/Expr.qll

```diff
@@ -283,19 +282,148 @@ class IsTypeExpr extends IsExpr {
   TypeAccess getTypeAccess() { result = typeAccess }
 }
 
-/**
- * An `is` pattern expression, for example `x is string s`.
- */
-class IsPatternExpr extends IsTypeExpr {
+/** An `is` pattern expression, for example `x is string s`. */
+class IsPatternExpr extends IsExpr {
   LocalVariableDeclExpr typeDecl;
 
-  IsPatternExpr() { typeDecl = this.getChild(2) }
+  IsPatternExpr() { typeDecl = this.getChild(2) and not this instanceof IsRecursivePatternExpr }
 
   /**
    * Gets the local variable declaration in this `is` pattern expression.
    * For example `string s` in `x is string s`.
    */
   LocalVariableDeclExpr getVariableDeclExpr() { result = typeDecl }
+
+  /**
+   * Gets the type being accessed in this `is` expression, for example `string`
+   * in `x is string`.
+   */
+  Type getCheckedType() { result = getTypeAccess().getTarget() }
+
+  /**
+   * Gets the type access in this `is` expression, for example `string` in
+   * `x is string`.
+   */
+  TypeAccess getTypeAccess() { result = this.getChild(1) }
+}
+
+/** An `is` recursive pattern expression, for example `x is string { Length: 5 } s`. */
+class IsRecursivePatternExpr extends IsExpr {
```

Comment:
Ok, so what is the QL class of
```
1) case { Length: 5 }:

2) foo is { Length: 5 }
```
1) Can't be a `TypeCase` because it isn't a type test. There is no variable declaration, and no type access. The recursive pattern is really a different beast to a `TypeCase`.

2) It can't be an `IsTypeExpr` because it isn't a type test, and it can't be an `IsPatternExpr` because there is no variable declared.

QL class hierarchy:

-  `IsExpr`
    - `IsConstantExpr`
    - `IsTypeExpr`
        - `IsPatternExpr`
            - `.getRecursivePattern()` ?? (Tom's proposal)
    - `IsRecursivePatternExpr` ?? (Calum's implementation)
- `CaseStmt`
    - `ConstCase`
    - `TypeCase`
        - `.getRecursivePattern()` ??
    - `RecursivePatternCase` ??

I do think it would be useful to be able to name a recursive pattern case/is by type, which is why it might make sense to keep `RecursivePatternCase` and `IsRecursivePatternExpr`. However, in the current implementation, these types are mutually exclusive with `TypeCase` and `IsPattern`, which I think is wrong.

In terms of control flow implementation, I might still want to keep `RecursivePatternCase` as a private class because it is so different to both a regular `CaseStmt` and it isn't a `TypeCase`.